testing/trino-testing/src/main/java/io/trino/testing/BaseConnectorTest.java

```diff
@@ -4859,6 +4949,90 @@ protected void verifyConcurrentAddColumnFailurePermissible(Exception e)
         throw new AssertionError("Unexpected concurrent add column failure", e);
     }
 
+    // Repeat test with invocationCount for better test coverage, since the tested aspect is inherently non-deterministic.
+    @Test(timeOut = 60_000, invocationCount = 4)
+    public void testCreateOrReplaceTableConcurrently()
+            throws Exception
+    {
+        if (!hasBehavior(SUPPORTS_CREATE_OR_REPLACE_TABLE)) {
+            // Already handled in testCreateOrReplaceTableWhenTableNotYetExists
+            return;
+        }
+
+        int threads = 4;
+        int numOfCreateOrReplaceStatements = 4;
+        int numOfReads = 16;
+        CyclicBarrier barrier = new CyclicBarrier(threads + 1);
+        ExecutorService executor = newFixedThreadPool(threads + 1);
+        List<Future<?>> futures = new ArrayList<>();
+        try (TestTable table = createTableWithOneIntegerColumn("test_create_or_replace")) {
+            String tableName = table.getName();
+
+            getQueryRunner().execute("CREATE OR REPLACE TABLE " + tableName + " AS SELECT 1 a");
+            assertThat(query("SELECT * FROM " + tableName)).matches("VALUES 1");
+
+            // One thread submits some CREATE OR REPLACE statements
+            futures.add(executor.submit(() -> {
+                barrier.await(30, SECONDS);
+                IntStream.range(0, numOfCreateOrReplaceStatements).forEach(index -> {
+                    try {
+                        String columnName = "col" + index;
+                        getQueryRunner().execute("CREATE OR REPLACE TABLE " + tableName + " AS SELECT 1 " + columnName);
+                    } catch (Exception e) {
+                        RuntimeException trinoException = getTrinoExceptionCause(e);
+                        try {
+                            verifyConcurrentCreateOrReplaceFailurePermissible(trinoException);
+                        } catch (Throwable verifyFailure) {
+                            if (verifyFailure != e) {
+                                verifyFailure.addSuppressed(e);
+                            }
+                            throw verifyFailure;
+                        }
+                    }
+                });
+                return null;
+            }));
+
+            // Other 4 threads continue try to read the same table, none of the reads should fail.
+            IntStream.range(0, threads)
+                    .forEach(threadNumber -> futures.add(executor.submit(() -> {
+                        barrier.await(30, SECONDS);
+                        IntStream.range(0, numOfReads).forEach(readIndex -> {
+                            try {
+                                assertThat(query("SELECT * FROM " + tableName)).matches("VALUES 1");
```

Comment:
```suggestion
            // One thread submits some CREATE OR REPLACE statements
            futures.add(executor.submit(() -> {
                barrier.await(30, SECONDS);
                IntStream.range(0, numOfCreateOrReplaceStatements).forEach(index -> {
                    try {
                        getQueryRunner().execute("CREATE OR REPLACE TABLE " + tableName + " AS SELECT * FROM (VALUES (1), (2)) AS t(a) ");
                    } catch (Exception e) {
                        RuntimeException trinoException = getTrinoExceptionCause(e);
                        try {
                            verifyConcurrentCreateOrReplaceFailurePermissible(trinoException);
                        } catch (Throwable verifyFailure) {
                            if (verifyFailure != e) {
                                verifyFailure.addSuppressed(e);
                            }
                            throw verifyFailure;
                        }
                    }
                });
                return null;
            }));

            // Other 4 threads continue try to read the same table, none of the reads should fail.
            IntStream.range(0, threads)
                    .forEach(threadNumber -> futures.add(executor.submit(() -> {
                        barrier.await(30, SECONDS);
                        IntStream.range(0, numOfReads).forEach(readIndex -> {
                            try {
                                MaterializedResult result = computeActual("SELECT * FROM " + tableName);
                                if (result.getRowCount() == 1) {
                                    assertEqualsIgnoreOrder(result.getMaterializedRows(), List.of(new MaterializedRow(List.of(1))));
                                }
                                else {
                                    assertEqualsIgnoreOrder(result.getMaterializedRows(), List.of(new MaterializedRow(List.of(1)), new MaterializedRow(List.of(2))));
                                }
                            }
                            catch (Exception e) {
                                RuntimeException trinoException = getTrinoExceptionCause(e);
                                try {
                                    verifyConcurrentCreateOrReplaceFailurePermissible(trinoException);
                                }
                                catch (Throwable verifyFailure) {
                                    if (verifyFailure != e) {
                                        verifyFailure.addSuppressed(e);
                                    }
                                    throw verifyFailure;
                                }
                            }
                        });
                        return null;
                    })));
            futures.forEach(Futures::getUnchecked);

            getQueryRunner().execute("CREATE OR REPLACE TABLE " + tableName + " AS SELECT * FROM (VALUES (1), (2), (3)) AS t(a)");
            assertThat(query("SELECT * FROM " + tableName)).matches("VALUES 1, 2, 3");
```

Reply:
This version gets a bit closer to real-world usage of CREATE OR REPLACE where the table structure does not necessarily change, but the content however gets changed.